### PR TITLE
tui: Enable TUI report mode to use option -f

### PR DESCRIPTION
**Pull request by qpakzk**
Hello.

Before this commit, the `tui` command provided the `-f` option only for the graph mode.
This commit allows the `tui` command to provide the `-f` option when adding the `--report` option together.

Fixed: #1278.

### CODE REVIEW

**Comment by qpakzk** [comment]
@honggyukim Thanks for your feedback. I updated your comment.
Also, I wrote the more specific commit message.

File: doc/uftrace-tui.md [hunk 1]
```diff
@@ -23,10 +23,16 @@ the initial data loading.
 TUI OPTIONS
 ===========
 -f *FIELD*, \--output-fields=*FIELD*
-:   Customize field in the output.  Possible values are: total, self and addr.
-    Multiple fields can be set by using comma.  Special field of 'none' can be
-    used (solely) to hide all fields.  Default is 'total'.
-    See `uftrace-graph`(1) for an explanation of fields.
+:   Customize fields in the output. This option basically indicates graph fields.
+    Possible values are total, self and addr. The default value is 'total'.
+    But if this option is combined with the --report option,
```
Comment (honggyukim) [inline doc/uftrace-tui.md:28]: I think "used" is better than "combined" here and don't need to use "the".
```suggestion
    But if this option is used with --report option,
```

File: doc/uftrace-tui.md [hunk 2]
```diff
@@ -23,10 +23,16 @@ the initial data loading.
 TUI OPTIONS
 ===========
 -f *FIELD*, \--output-fields=*FIELD*
-:   Customize field in the output.  Possible values are: total, self and addr.
-    Multiple fields can be set by using comma.  Special field of 'none' can be
-    used (solely) to hide all fields.  Default is 'total'.
-    See `uftrace-graph`(1) for an explanation of fields.
+:   Customize fields in the output. This option basically indicates graph fields.
```
Comment (honggyukim) [inline doc/uftrace-tui.md:26]: By convention in uftrace docs, we give 2 white spaces after the period.
```suggestion
:   Customize fields in the output.  This option basically indicates graph fields.
```
Please apply the same for the rest of other sentences.

File: cmds/tui.c [hunk 3]
```diff
@@ -317,45 +317,46 @@ static struct display_field *graph_field_table[] = {
...
-	.list    = LIST_HEAD_INIT(report_field_##_func.list)		\
+#define REPORT_FIELD_STRUCT(_id, _name, _alias, _func, _header, _length)	\
+static struct display_field report_field_##_func = {				\
+	.id      = _id,								\
+	.name    = #_name,							\
+	.alias   = #_alias,							\
+	.header  = _header,							\
+	.length  = _length,							\
+	.print   = print_report_##_func,					\
+	.list    = LIST_HEAD_INIT(report_field_##_func.list)			\
```
Comment (honggyukim) [inline cmds/tui.c:328]: We don't need to add `_alias` here. Please revert this change.

File: cmds/tui.c [hunk 4]
```diff
@@ -317,45 +317,46 @@ static struct display_field *graph_field_table[] = {
...
+
+REPORT_FIELD_TIME(REPORT_F_TOTAL_TIME, total, total, total.sum, total, "TOTAL TIME");
+REPORT_FIELD_TIME(REPORT_F_TOTAL_TIME_AVG, total-avg, total-avg, total.avg, total_avg, "TOTAL AVG");
+REPORT_FIELD_TIME(REPORT_F_TOTAL_TIME_MIN, total-min, total-min, total.min, total_min, "TOTAL MIN");
+REPORT_FIELD_TIME(REPORT_F_TOTAL_TIME_MAX, total-max, total-max, total.max, total_max, "TOTAL MAX");
+REPORT_FIELD_TIME(REPORT_F_SELF_TIME, self, self, self.sum, self, "SELF TIME");
+REPORT_FIELD_TIME(REPORT_F_SELF_TIME_AVG, self-avg, self-avg, self.avg, self_avg, "SELF AVG");
+REPORT_FIELD_TIME(REPORT_F_SELF_TIME_MIN, self-min, self-min, self.min, self_min, "SELF MIN");
+REPORT_FIELD_TIME(REPORT_F_SELF_TIME_MAX, self-max, self-max, self.max, self_max, "SELF MAX");
+REPORT_FIELD_CALL(REPORT_F_CALL, call, call, call, call, "CALL");
```
Comment (honggyukim) [inline cmds/tui.c:359]: By removing `_alias` change, we can remove all these changes in the diff.

**Review by honggyukim — COMMENTED** [review]
One of the unit test fails with this change.
```
$ make unittest
  TEST     test_unit
Running 69 test cases
    ...
[025] field_setup_list              : FAIL
```
Please see the test at https://github.com/namhyung/uftrace/blob/master/utils/field.c#L224-L239.

**Comment by qpakzk** [comment]
@honggyukim Thanks for your review.

```sh
$ uftrace tui -f +call
uftrace: Unknown field name 'call'
uftrace:   Possible fields are: total-time self-time address
```

The possible values and the error message values for `-f` option are different in TUI graph mode.
Users might make a mistake of using `total-time`, `self-time` or `address` when adding values for `-f` option.
How about adding `.alias` and showing `.alias` instead?

```diff
diff --git a/utils/field.c b/utils/field.c
index f19da6ba..94666969 100644
--- a/utils/field.c
+++ b/utils/field.c
@@ -165,7 +165,7 @@ void setup_field(struct list_head *output_fields, struct opts *opts,
                        pr_out("uftrace: Unknown field name '%s'\n", p);
                        pr_out("uftrace:   Possible fields are:");
                        for (i = 0; i < field_table_size; i++)
-                               pr_out(" %s", field_table[i]->name);
+                               pr_out(" %s", field_table[i]->alias);
                        pr_out("\n");
                        exit(1);
                }
```

**Comment by qpakzk** [comment]
@honggyukim Thanks for your review. I misunderstood something.
I applied your advice to my commit.
Also, I updated tests.

**Comment by qpakzk** [comment]
I found a problem when printing an error message.

When my shell is on the bottom, it works fine.
Screen Shot 2021-06-19 at 19 00 46

However, when my shell isn't on the bottom, my shell overlaps the error message.
Screen Shot 2021-06-19 at 19 01 04

**Review by honggyukim — COMMENTED** [review]
Sorry for the late response and thanks for the update again. Could you split this patch into two patches? Otherwise looks good to me.
1. field: Keep the predefined field order when using -f option
2. tui: Allow tui report mode to use option -f

The field keeps ordering not only in report command but also the other commands such as graph and replay commands.

**Comment by qpakzk** [comment]
@honggyukim Thanks for your feedback.
I updated your requirement.

**Review by honggyukim — APPROVED** [review]
LGTM. Thanks for your work!

**Comment by honggyukim** [comment]
@qpakzk We can also use `-s`/`--sort` option when `--report` option is used with this change. If you want to work more about TUI, then this can be a possible next work item. :)

**Comment by qpakzk** [comment]
@honggyukim I'm gonna try it!

File: utils/field.c [hunk 5]
```diff
@@ -129,11 +130,17 @@ void setup_field(struct list_head *output_fields, struct opts *opts,
 	if (!strcmp(opts->fields, "none"))
 		return;
 
+	field_flags = calloc(field_table_size, sizeof(bool));
```
Comment (namhyung) [inline utils/field.c:133]: Please use `xcalloc` to check error.  Also I prefer using `sizeof(*field_flags)` than the type name (bool) because it's easier to deal with future (type) changes.

**Review by namhyung — COMMENTED** [review]
I have a nitpick, but otherwise looks ok.

**Review by namhyung — APPROVED** [review]
LGTM
